### ISSUE TITLE
`assert outer_fence is not None` on migration adding link to new type

### ISSUE THREAD

**Issue by jackfischer** [edgedb/edgedb#6348]
- EdgeDB Version: 3.4, **fixed in 4.0-beta.1+bc03568**
- EdgeDB CLI Version: 3.5.0+e1ad387
- OS Version: macos 13

Steps to Reproduce:

1. Start with the following base schema, with Organization/User/Old types
```
CREATE MIGRATION m1dukh4kdrpf2sd45ft3pgxfgzhfpjdtknjy6whmhdhovaapdgkssq
    ONTO initial
{
  CREATE TYPE default::Organization;
  CREATE TYPE default::User {
      CREATE REQUIRED SINGLE LINK organization: default::Organization;
  };
  CREATE TYPE default::Old {
      CREATE REQUIRED SINGLE LINK user: default::User;
  };
};
```

2. Perform a migration to add a type `New` pointed to from `Old` with a field back populated from `Old`
```
CREATE MIGRATION m1t2elmv2z2zrtz7qzwrwxdj3ml4yijissv5ek2fk4vi7kytolnoza
    ONTO m1dukh4kdrpf2sd45ft3pgxfgzhfpjdtknjy6whmhdhovaapdgkssq
{
  CREATE TYPE default::New {
      CREATE REQUIRED SINGLE LINK organization: default::Organization;
  };
  ALTER TYPE default::Old {
      CREATE REQUIRED SINGLE LINK new: default::New {
          SET REQUIRED USING (INSERT
              default::New
              {
                  organization := .user.organization
              });
      };
  };
};
```

Recieve the following error,
```
edgedb error: InternalServerError: None
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      Traceback (most recent call last):
        File "/Users/jack/Library/Application Support/edgedb/portable/3.4/lib/python3.11/site-packages/edb/pgsql/compiler/__init__.py", line 135, in compile_ir_to_sql_tree
          qtree = dispatch.compile(ir_expr, ctx=ctx)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/jack/Library/Application Support/edgedb/portable/3.4/lib/python3.11/functools.py", line 909, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/jack/Library/Application Support/edgedb/portable/3.4/lib/python3.11/site-packages/edb/pgsql/compiler/expr.py", line 60, in compile_Set
          _compile_set_impl(ir_set, ctx=ctx)

...

        File "/Users/jack/Library/Application Support/edgedb/portable/3.4/lib/python3.11/site-packages/edb/pgsql/compiler/relgen.py", line 1309, in process_set_as_subquery
          assert outer_fence is not None
                 ^^^^^^^^^^^^^^^^^^^^^^^
```          

Schema:
```
module default {

    type Organization {}
    type User {
        required single organization: Organization;
    }


    type New {
        required single organization: Organization;
    }

    type Old {
        required single new: New;
        required single user: User;
    }

}
```

**Comment by KaelWD** [comment]
I just got this too but when modifying a link

```
edgedb error: InternalServerError: None
Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
Server traceback:
    Traceback (most recent call last):
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/compiler/__init__.py", line 135, in compile_ir_to_sql_tree
        qtree = dispatch.compile(ir_expr, ctx=ctx)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/functools.py", line 909, in wrapper
        return dispatch(args[0].__class__)(*args, **kw)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      ...
      
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/compiler/relgen.py", line 222, in get_set_rvar
        rvars = _get_set_rvar(ir_set, ctx=subctx)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/compiler/relgen.py", line 348, in _get_set_rvar
        return process_set_as_subquery(ir_set, ctx=ctx)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/compiler/relgen.py", line 1309, in process_set_as_subquery
        assert outer_fence is not None
               ^^^^^^^^^^^^^^^^^^^^^^^
    AssertionError
    
    The above exception was the direct cause of the following exception:
    
    Traceback (most recent call last):
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/server/compiler_pool/worker.py", line 186, in compile_in_tx
        units, cstate = COMPILER.compile_in_tx(cstate, *args, **kwargs)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      ...
      
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/delta.py", line 5657, in apply
        self._alter_pointer_optionality(
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/delta.py", line 4172, in _alter_pointer_optionality
        (conv_expr_ctes, _) = self._compile_conversion_expr(
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/delta.py", line 4698, in _compile_conversion_expr
        sql_tree, env, _ = compiler.compile_ir_to_sql_tree(
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/x86_64-linux-gnu/edgedb-server-3/lib/python3.11/site-packages/edb/pgsql/compiler/__init__.py", line 152, in compile_ir_to_sql_tree
        raise errors.InternalServerError(*args) from e
    edb.errors.InternalServerError: None
edgedb error: error in one of the migrations
```

```
CREATE MIGRATION m1buhfv466vzjhawm4kbdermgcbf67hytpskakcvb5mkmez3vvbvua
    ONTO m1j32sene3xpj2kma4dqgbk2awwj2uiyufjouduzhhoimeb2zkh6cq
{
  ALTER TYPE default::Identity {
      CREATE REQUIRED MULTI PROPERTY emails: std::str {
          SET REQUIRED USING (.user.emails);
      };
      CREATE REQUIRED PROPERTY updatedAt: std::datetime {
          SET default := (std::datetime_of_transaction());
          CREATE REWRITE
              UPDATE 
              USING (std::datetime_of_transaction());
      };
  };
  ALTER TYPE default::Session {
      ALTER PROPERTY lastActive {
          SET default := (std::datetime_of_transaction());
          DROP REWRITE
              INSERT ;
          };
      };
  ALTER TYPE default::Sponsorship {
      ALTER PROPERTY amount {
          SET TYPE std::int32 USING (<std::int32>.amount);
      };
      ALTER PROPERTY platform {
          SET readonly := true;
      };
      ALTER PROPERTY target {
          RESET OPTIONALITY;
      };
      CREATE REQUIRED PROPERTY updatedAt: std::datetime {
          SET default := (std::datetime_of_transaction());
          CREATE REWRITE
              UPDATE 
              USING (std::datetime_of_transaction());
      };
  };
  ALTER TYPE default::User {
      ALTER LINK identities {
          ON TARGET DELETE ALLOW;
      };
      DROP PROPERTY emails;
  };
};
```

**Comment by msullivan** [comment]
Fixed by #6267. Looks like it works in 3.5.
